File: packages/app-game-domain/src/app-game-policy-preview-handoff.ts
import {
  type Infer,
  Schema,
  withParser,
  brandedNonEmptyStringSchema,
  NonEmptyStringSchema
} from '@ocentra-parent/schema-domain/effect';
import {
  AppGamePolicyCompiledDecisionSchema,
  AppGamePolicyCompiledDecisionIdSchema,
  AppGamePolicyCompileRequestIdSchema,
  AppGamePolicyCompilerOutcomeStateSchema,
  AppGamePolicyCompilerRejectionReasonSchema,
  AppGamePolicyTargetKindSchema,
} from './app-game-policy-target-compiler';
import {
  AppGamePolicyPreviewStatus,
  AppGamePolicyPreviewTargetDomain,
  AppGamePolicyPreviewNoRuntimeClaimFlags,
  AppGamePolicyPreviewNoRuntimeClaimStates,
  countAppGamePolicyPreviewReadModelRows,
  appGamePolicyPreviewReadModelCountsMatchRows,
  appGamePolicyPreviewRowHasNoRuntimeClaims,
  appGamePolicyPreviewRowHasProofRefs,
  appGamePolicyPreviewRowIsDryRunOnly,
  appGamePolicyPreviewStatusForOutcome,
  appGamePolicyPreviewStatusMatchesOutcome,
  appGamePolicyPreviewTargetDomainForKind,
} from './app-game-policy-preview-handoff-rules';
import {
  PolicyActionSchema,
  PolicyDecisionHandoffStateSchema,
  PolicyDecisionIdSchema,
  PolicyRuleIdSchema,
  PolicyTargetSchema,
} from '@ocentra-parent/policy-domain/policy';
import { ParentDeviceReferenceSchema, ParentEvidenceReferenceSchema } from '@ocentra-parent/family-domain/references';
import {
  ParentContractSchemaVersionSchema,
  ParentPolicyVersionSchema,
  ParentTimestampSchema,
} from '@ocentra-parent/schema-domain/family-reference-primitives';

export const AppGamePolicyPreviewHandoffIdSchema = brandedNonEmptyStringSchema('AppGamePolicyPreviewHandoffId');
export const AppGamePolicyPreviewHandoffRowIdSchema = brandedNonEmptyStringSchema('AppGamePolicyPreviewHandoffRowId');
export const AppGamePolicyPreviewHandoffSourceContractRefSchema = brandedNonEmptyStringSchema('AppGamePolicyPreviewHandoffSourceContractRef');
export const AppGamePolicyPreviewHandoffRuntimeClaimStateSchema = withParser(Schema.Literal('not-claimed'));
export const AppGamePolicyPreviewHandoffAdapterDispatchStateSchema = withParser(Schema.Literal('not-dispatched'));

export const AppGamePolicyPreviewTargetDomainSchema = withParser(
  Schema.Literal(...Object.values(AppGamePolicyPreviewTargetDomain))
);
export const AppGamePolicyPreviewStatusSchema = withParser(
  Schema.Literal(...Object.values(AppGamePolicyPreviewStatus))
);

export const AppGamePolicyPreviewHandoffOptionsSchema = withParser(
  Schema.Struct({
    schemaVersion: ParentContractSchemaVersionSchema,
    handoffId: AppGamePolicyPreviewHandoffIdSchema,
    generatedAt: ParentTimestampSchema,
    sourceContractRefs: Schema.Array(AppGamePolicyPreviewHandoffSourceContractRefSchema),
  }).pipe(
    Schema.filter(
      (options) =>
        options.sourceContractRefs.length > 0 ||
        'Expected app/game policy preview handoff options to cite source contracts'
    )
  )
);

const AppGamePolicyPreviewHandoffRowBaseSchema = Schema.Struct({
  schemaVersion: ParentContractSchemaVersionSchema,
  rowId: AppGamePolicyPreviewHandoffRowIdSchema,
  targetDomain: AppGamePolicyPreviewTargetDomainSchema,
  sourceCompiledDecisionId: AppGamePolicyCompiledDecisionIdSchema,
  sourceCompileRequestId: AppGamePolicyCompileRequestIdSchema,
  sourceTargetKind: AppGamePolicyTargetKindSchema,
  device: ParentDeviceReferenceSchema,
  policyVersion: ParentPolicyVersionSchema,
  policyTarget: PolicyTargetSchema,
  policyDecisionId: PolicyDecisionIdSchema,
  policyAction: PolicyActionSchema,
  outcomeState: AppGamePolicyCompilerOutcomeStateSchema,
  previewStatus: AppGamePolicyPreviewStatusSchema,
  rejectionReason: AppGamePolicyCompilerRejectionReasonSchema,
  ruleRefs: Schema.Array(PolicyRuleIdSchema),
  evidenceReferences: Schema.Array(ParentEvidenceReferenceSchema),
  capabilityRefs: Schema.Array(NonEmptyStringSchema),
  authorityRefs: Schema.Array(NonEmptyStringSchema),
  auditRefs: Schema.Array(NonEmptyStringSchema),
  dryRun: Schema.Boolean,
  enforcementHandoffState: PolicyDecisionHandoffStateSchema,
  policyEvaluatorRuntimeClaimState: AppGamePolicyPreviewHandoffRuntimeClaimStateSchema,
  timerRuntimeClaimState: AppGamePolicyPreviewHandoffRuntimeClaimStateSchema,
  adapterDispatchState: AppGamePolicyPreviewHandoffAdapterDispatchStateSchema,
  childDeliveryClaimState: AppGamePolicyPreviewHandoffRuntimeClaimStateSchema,
  platformEnforcementClaimState: AppGamePolicyPreviewHandoffRuntimeClaimStateSchema,
  policyEvaluatorRuntimeClaimed: Schema.Boolean,
  timerRuntimeClaimed: Schema.Boolean,
  adapterDispatchClaimed: Schema.Boolean,
  childDeliveryClaimed: Schema.Boolean,
  platformEnforcementClaimed: Schema.Boolean,
  generatedAt: ParentTimestampSchema,
});

export const AppGamePolicyPreviewHandoffRowSchema = withParser(
  AppGamePolicyPreviewHandoffRowBaseSchema.pipe(
    Schema.filter(
      (row) =>
        appGamePolicyPreviewStatusMatchesOutcome(row) ||
        'Expected app/game policy preview rows to preserve compiler outcome state'
    )
  )
    .pipe(
      Schema.filter(
        (row) =>
          appGamePolicyPreviewRowIsDryRunOnly(row) ||
          'Expected app/game policy preview rows to remain dry-run with disabled enforcement handoff'
      )
    )
    .pipe(
      Schema.filter(
        (row) =>
          appGamePolicyPreviewRowHasNoRuntimeClaims(row) ||
          'Expected app/game policy preview rows to avoid evaluator, timer, adapter, child-delivery, and platform enforcement claims'
      )
    )
    .pipe(
      Schema.filter(
        (row) =>
          appGamePolicyPreviewRowHasProofRefs(row) ||
          'Expected app/game policy preview rows to retain evidence, rule, capability, and audit refs'
      )
    )
);

const AppGamePolicyPreviewHandoffReadModelBaseSchema = Schema.Struct({
  schemaVersion: ParentContractSchemaVersionSchema,
  handoffId: AppGamePolicyPreviewHandoffIdSchema,
  generatedAt: ParentTimestampSchema,
  sourceContractRefs: Schema.Array(AppGamePolicyPreviewHandoffSourceContractRefSchema),
  rows: Schema.Array(AppGamePolicyPreviewHandoffRowSchema),
  nativeAppRowCount: Schema.Number,
  nativeGameRowCount: Schema.Number,
  previewReadyCount: Schema.Number,
  manualRequiredCount: Schema.Number,
  rejectedCount: Schema.Number,
  policyEvaluatorRuntimeClaimed: Schema.Boolean,
  timerRuntimeClaimed: Schema.Boolean,
  adapterDispatchClaimed: Schema.Boolean,
  childDeliveryClaimed: Schema.Boolean,
  platformEnforcementClaimed: Schema.Boolean,
});

export const AppGamePolicyPreviewHandoffReadModelSchema = withParser(
  AppGamePolicyPreviewHandoffReadModelBaseSchema.pipe(
    Schema.filter(
      (readModel) =>
        appGamePolicyPreviewReadModelCountsMatchRows(readModel) ||
        'Expected app/game policy preview handoff counts to match rows'
    )
  )
    .pipe(
      Schema.filter(
        (readModel) =>
          new Set(readModel.rows.map((row) => row.rowId)).size === readModel.rows.length ||
          'Expected app/game policy preview handoff row ids to be unique'
      )
    )
    .pipe(
      Schema.filter(
        (readModel) =>
          (!readModel.policyEvaluatorRuntimeClaimed &&
            !readModel.timerRuntimeClaimed &&
            !readModel.adapterDispatchClaimed &&
            !readModel.childDeliveryClaimed &&
            !readModel.platformEnforcementClaimed) ||
          'Expected app/game policy preview handoff read model to avoid runtime and enforcement claims'
      )
    )
);

export type AppGamePolicyPreviewHandoffOptions = Infer<typeof AppGamePolicyPreviewHandoffOptionsSchema>;
export type AppGamePolicyPreviewHandoffRow = Infer<typeof AppGamePolicyPreviewHandoffRowSchema>;
export type AppGamePolicyPreviewHandoffReadModel = Infer<typeof AppGamePolicyPreviewHandoffReadModelSchema>;
export type AppGamePolicyPreviewTargetDomain = Infer<typeof AppGamePolicyPreviewTargetDomainSchema>;
export type AppGamePolicyPreviewStatus = Infer<typeof AppGamePolicyPreviewStatusSchema>;

export function buildAppGamePolicyPreviewHandoffReadModel(
  optionsInput: unknown,
  decisionsInput: readonly unknown[]
): AppGamePolicyPreviewHandoffReadModel {
  const options = AppGamePolicyPreviewHandoffOptionsSchema.parse(optionsInput);
  const rows = decisionsInput.map((decision) => buildAppGamePolicyPreviewHandoffRow(options, decision));
  return AppGamePolicyPreviewHandoffReadModelSchema.parse({
    schemaVersion: options.schemaVersion,
    handoffId: options.handoffId,
    generatedAt: options.generatedAt,
    sourceContractRefs: options.sourceContractRefs,
    rows,
    ...countAppGamePolicyPreviewReadModelRows(rows),
    ...AppGamePolicyPreviewNoRuntimeClaimFlags,
  });
}

export function buildAppGamePolicyPreviewHandoffRow(
  optionsInput: unknown,
  decisionInput: unknown
): AppGamePolicyPreviewHandoffRow {
  const options = AppGamePolicyPreviewHandoffOptionsSchema.parse(optionsInput);
  const decision = AppGamePolicyCompiledDecisionSchema.parse(decisionInput);
  return AppGamePolicyPreviewHandoffRowSchema.parse({
    schemaVersion: decision.schemaVersion,
    rowId: `${decision.compiledDecisionId}:preview`,
    targetDomain: appGamePolicyPreviewTargetDomainForKind(decision.request.target.targetKind),
    sourceCompiledDecisionId: decision.compiledDecisionId,
    sourceCompileRequestId: decision.request.compileRequestId,
    sourceTargetKind: decision.request.target.targetKind,
    device: decision.request.device,
    policyVersion: decision.request.policyVersion,
    policyTarget: decision.policyTarget,
    policyDecisionId: decision.policyDecision.decisionId,
    policyAction: decision.policyDecision.action,
    outcomeState: decision.outcomeState,
    previewStatus: appGamePolicyPreviewStatusForOutcome(decision.outcomeState),
    rejectionReason: decision.rejectionReason,
    ruleRefs: decision.policyDecision.ruleIds,
    evidenceReferences: decision.policyDecision.evidenceReferences,
    capabilityRefs: decision.capabilityRefs,
    authorityRefs: decision.authorityRefs,
    auditRefs: decision.auditRefs,
    dryRun: decision.policyDecision.dryRun,
    enforcementHandoffState: decision.policyDecision.enforcementHandoffState,
    ...AppGamePolicyPreviewNoRuntimeClaimStates,
    ...AppGamePolicyPreviewNoRuntimeClaimFlags,
    generatedAt: options.generatedAt,
  });
}

export const decodeAppGamePolicyPreviewHandoffReadModel = Schema.decodeUnknownSync(
  AppGamePolicyPreviewHandoffReadModelSchema
);

export { AppGamePolicyPreviewStatus, AppGamePolicyPreviewTargetDomain };

File: crates/app-game-core/tests/contract/app_game_policy_preview_handoff.rs
use ocentra_app_game_core::app_game_policy_preview_handoff::{
    app_game_policy_preview_handoff_typescript, build_app_game_policy_preview_handoff_read_model,
    build_app_game_policy_preview_handoff_row, AppGamePolicyPreviewCompiledDecision,
    AppGamePolicyPreviewHandoffOptions,
};

#[test]
fn app_game_policy_preview_handoff_builds_native_app_preview_rows_without_runtime_claims() {
    let row =
        build_app_game_policy_preview_handoff_row(&preview_options(), &app_compiled_decision());

    assert_eq!(row.target_domain, "native-app");
    assert_eq!(row.preview_status, "preview-ready");
    assert_eq!(row.policy_action, "time-limit");
    assert_eq!(
        row.source_compiled_decision_id,
        "compiled-decision-preview-app"
    );
    assert_eq!(
        row.evidence_references,
        vec!["app-game-policy-preview-evidence-1"]
    );
    assert_eq!(row.rule_refs, vec!["policy-rule-app-game-preview-1"]);
    assert_eq!(row.capability_refs, vec!["capability-preview-1"]);
    assert_eq!(row.audit_refs, vec!["audit-preview-1"]);
    assert!(row.dry_run);
    assert_eq!(row.enforcement_handoff_state, "disabled");
    assert!(!row.policy_evaluator_runtime_claimed);
    assert!(!row.timer_runtime_claimed);
    assert!(!row.adapter_dispatch_claimed);
    assert!(!row.child_delivery_claimed);
    assert!(!row.platform_enforcement_claimed);
}

#[test]
fn app_game_policy_preview_handoff_keeps_game_manual_rows_manual_required() {
    let row =
        build_app_game_policy_preview_handoff_row(&preview_options(), &game_manual_decision());

    assert_eq!(row.target_domain, "native-game");
    assert_eq!(row.preview_status, "manual-required");
    assert_eq!(row.rejection_reason, "block-launch-manual-required");
    assert_eq!(row.policy_action, "block");
    assert_eq!(row.adapter_dispatch_state, "not-dispatched");
    assert!(!row.adapter_dispatch_claimed);
    assert!(!row.platform_enforcement_claimed);
}

#[test]
fn app_game_policy_preview_handoff_read_model_counts_rows_by_domain_and_status() {
    let read_model = build_app_game_policy_preview_handoff_read_model(
        &preview_options(),
        &[app_compiled_decision(), game_manual_decision()],
    );

    assert_eq!(read_model.native_app_row_count, 1);
    assert_eq!(read_model.native_game_row_count, 1);
    assert_eq!(read_model.preview_ready_count, 1);
    assert_eq!(read_model.manual_required_count, 1);
    assert_eq!(read_model.rejected_count, 0);
    assert_eq!(
        read_model
            .rows
            .iter()
            .map(|row| row.row_id.as_str())
            .collect::<Vec<_>>(),
        vec![
            "compiled-decision-preview-app:preview",
            "compiled-decision-preview-game-manual:preview",
        ]
    );
    assert_eq!(
        read_model.source_contract_refs,
        vec![
            "app-game-policy-target-compiler",
            "docs/expectations/policy.md",
            "docs/expectations/enforcement.md",
        ]
    );
    assert!(!read_model.policy_evaluator_runtime_claimed);
    assert!(!read_model.adapter_dispatch_claimed);
}

#[test]
fn generated_app_game_policy_preview_handoff_helper_stays_checked_in() {
    let checked_in = include_str!(
        "../../../../packages/schema-domain/src/generated-app-game-policy-preview-handoff-values.ts"
    );
    let generated = app_game_policy_preview_handoff_typescript();

    assert_eq!(checked_in, generated);
}

fn preview_options() -> AppGamePolicyPreviewHandoffOptions {
    AppGamePolicyPreviewHandoffOptions {
        schema_version: "v0.6".to_string(),
        handoff_id: "app-game-policy-preview-handoff-proof".to_string(),
        generated_at: "2026-06-05T14:45:00Z".to_string(),
        source_contract_refs: vec![
            "app-game-policy-target-compiler".to_string(),
            "docs/expectations/policy.md".to_string(),
            "docs/expectations/enforcement.md".to_string(),
        ],
    }
}

fn app_compiled_decision() -> AppGamePolicyPreviewCompiledDecision {
    AppGamePolicyPreviewCompiledDecision {
        schema_version: "v0.6".to_string(),
        compiled_decision_id: "compiled-decision-preview-app".to_string(),
        compile_request_id: "compile-request-preview-app".to_string(),
        target_kind: "specific-app".to_string(),
        policy_decision_id: "policy-decision-preview-app".to_string(),
        policy_action: "time-limit".to_string(),
        outcome_state: "dry-run-ready".to_string(),
        rejection_reason: "none".to_string(),
        rule_refs: vec!["policy-rule-app-game-preview-1".to_string()],
        evidence_references: vec!["app-game-policy-preview-evidence-1".to_string()],
        capability_refs: vec!["capability-preview-1".to_string()],
        authority_refs: vec!["authority-preview-1".to_string()],
        audit_refs: vec!["audit-preview-1".to_string()],
        dry_run: true,
        enforcement_handoff_state: "disabled".to_string(),
    }
}

fn game_manual_decision() -> AppGamePolicyPreviewCompiledDecision {
    AppGamePolicyPreviewCompiledDecision {
        schema_version: "v0.6".to_string(),
        compiled_decision_id: "compiled-decision-preview-game-manual".to_string(),
        compile_request_id: "compile-request-preview-game".to_string(),
        target_kind: "specific-game".to_string(),
        policy_decision_id: "policy-decision-preview-game".to_string(),
        policy_action: "block".to_string(),
        outcome_state: "manual-required".to_string(),
        rejection_reason: "block-launch-manual-required".to_string(),
        rule_refs: vec!["policy-rule-app-game-preview-1".to_string()],
        evidence_references: vec!["app-game-policy-preview-evidence-1".to_string()],
        capability_refs: vec!["capability-preview-game-manual".to_string()],
        authority_refs: vec!["authority-preview-game-manual".to_string()],
        audit_refs: vec!["audit-preview-game-manual".to_string()],
        dry_run: true,
        enforcement_handoff_state: "disabled".to_string(),
    }
}
